File: backend/app/translation/languages.py
"""Centralized language registry for the multilingual architecture
(Milestone 1).

This is the SINGLE source of truth for which languages the application
knows about. It is mirrored field-for-field by
frontend/lib/i18n/languages.ts -- kept in sync by a parity test
(tests/test_language_registry.py). Never scatter language checks
(`if code == "hi": ...`) through the application; add or adjust a
language here instead.

Two independent flags, deliberately not conflated:

  * `enabled` -- whether the language has a static UI dictionary and can
    be chosen in the language selector. Milestone 1 enables exactly
    en/hi/bn; the other ten languages are registered (so adding one
    later is configuration-only, never an architecture change) but not
    yet selectable.

  * `capabilities` -- whether the REAL, configured BHASHINI pipeline has
    been verified to support that language for that task. As of this
    milestone no BHASHINI credentials/pipeline have been configured or
    verified against the live API (see
    docs/architecture/multilingual-architecture.md), so every
    capability below is False for every language, including en/hi/bn.
    `TRANSLATION_PROVIDER=mock` is what makes dynamic-content
    translation work in this milestone regardless of these flags --
    flip a capability to True only after verifying that exact
    (language, task) pair against the actual BHASHINI account/pipeline,
    never speculatively.
"""

from dataclasses import dataclass, field
from typing import Literal

LanguageCode = Literal[
    "en", "hi", "bn", "ta", "te", "mr", "gu", "kn", "ml", "or", "pa", "as", "ur"
]

Direction = Literal["ltr", "rtl"]


@dataclass(frozen=True)
class LanguageCapabilities:
    """Real BHASHINI infrastructure support for one language -- see this
    module's own docstring for why every field defaults to False."""

    translation: bool = False
    speech_to_text: bool = False
    text_to_speech: bool = False
    transliteration: bool = False


@dataclass(frozen=True)
class LanguageDefinition:
    code: LanguageCode
    name: str
    native_name: str
    locale: str
    direction: Direction
    enabled: bool
    capabilities: LanguageCapabilities = field(default_factory=LanguageCapabilities)


LANGUAGES: tuple[LanguageDefinition, ...] = (
    LanguageDefinition("en", "English", "English", "en-IN", "ltr", enabled=True),
    LanguageDefinition("hi", "Hindi", "हिन्दी", "hi-IN", "ltr", enabled=True),
    LanguageDefinition("bn", "Bengali", "বাংলা", "bn-IN", "ltr", enabled=True),
    LanguageDefinition("ta", "Tamil", "தமிழ்", "ta-IN", "ltr", enabled=False),
    LanguageDefinition("te", "Telugu", "తెలుగు", "te-IN", "ltr", enabled=False),
    LanguageDefinition("mr", "Marathi", "मराठी", "mr-IN", "ltr", enabled=False),
    LanguageDefinition("gu", "Gujarati", "ગુજરાતી", "gu-IN", "ltr", enabled=False),
    LanguageDefinition("kn", "Kannada", "ಕನ್ನಡ", "kn-IN", "ltr", enabled=False),
    LanguageDefinition("ml", "Malayalam", "മലയാളം", "ml-IN", "ltr", enabled=False),
    LanguageDefinition("or", "Odia", "ଓଡ଼ିଆ", "or-IN", "ltr", enabled=False),
    LanguageDefinition("pa", "Punjabi", "ਪੰਜਾਬੀ", "pa-IN", "ltr", enabled=False),
    LanguageDefinition("as", "Assamese", "অসমীয়া", "as-IN", "ltr", enabled=False),
    LanguageDefinition("ur", "Urdu", "اردو", "ur-IN", "rtl", enabled=False),
)

DEFAULT_LANGUAGE: LanguageCode = "en"

_BY_CODE: dict[str, LanguageDefinition] = {lang.code: lang for lang in LANGUAGES}


def get_language(code: str) -> LanguageDefinition | None:
    """One language definition by code, or None if unregistered. Never
    raises -- callers decide whether an unknown code means 'fall back to
    English' or 'reject the request'."""
    return _BY_CODE.get(code)


def is_supported(code: str) -> bool:
    """True only for a language that is both registered AND enabled --
    the check a caller almost always wants (e.g. validating a
    translation request's targetLanguage)."""
    lang = _BY_CODE.get(code)
    return lang is not None and lang.enabled


def enabled_languages() -> list[LanguageDefinition]:
    """Every language currently selectable in the UI (Milestone 1:
    en/hi/bn), in registry order."""
    return [lang for lang in LANGUAGES if lang.enabled]


def supported_codes() -> tuple[str, ...]:
    """Every registered code, enabled or not -- e.g. for validating a
    code against "is this a real language we know about at all"."""
    return tuple(lang.code for lang in LANGUAGES)

File: frontend/lib/i18n/languages.ts
/**
 * Centralized language registry for the multilingual architecture
 * (Milestone 1).
 *
 * This is the SINGLE source of truth for which languages the frontend
 * knows about. It mirrors backend/app/translation/languages.py
 * field-for-field -- kept in sync by a backend parity test
 * (backend/tests/test_language_registry.py). Never scatter language
 * checks (`if (code === "hi") ...`) through components; add or adjust a
 * language here instead.
 *
 * Two independent flags, deliberately not conflated:
 *
 *   - `enabled` -- whether the language has a static UI dictionary and
 *     can be chosen in the language selector. Milestone 1 enables
 *     exactly en/hi/bn; the other ten languages are registered (so
 *     adding one later is configuration-only, never an architecture
 *     change) but not yet selectable.
 *
 *   - `capabilities` -- whether the REAL, configured BHASHINI pipeline
 *     has been verified to support that language for that task. As of
 *     this milestone no BHASHINI credentials/pipeline have been
 *     configured or verified against the live API (see
 *     docs/architecture/multilingual-architecture.md), so every
 *     capability below is false for every language, including en/hi/bn.
 *     The mock translation provider is what makes dynamic-content
 *     translation work in this milestone regardless of these flags --
 *     flip a capability to true only after verifying that exact
 *     (language, task) pair against the actual BHASHINI account/
 *     pipeline, never speculatively.
 */

export type LanguageCode =
  | "en"
  | "hi"
  | "bn"
  | "ta"
  | "te"
  | "mr"
  | "gu"
  | "kn"
  | "ml"
  | "or"
  | "pa"
  | "as"
  | "ur";

export type Direction = "ltr" | "rtl";

export interface LanguageCapabilities {
  translation: boolean;
  speechToText: boolean;
  textToSpeech: boolean;
  transliteration: boolean;
}

export interface LanguageDefinition {
  code: LanguageCode;
  name: string;
  nativeName: string;
  locale: string;
  direction: Direction;
  enabled: boolean;
  capabilities: LanguageCapabilities;
}

const NO_CAPABILITIES: LanguageCapabilities = {
  translation: false,
  speechToText: false,
  textToSpeech: false,
  transliteration: false,
};

export const LANGUAGES: readonly LanguageDefinition[] = [
  { code: "en", name: "English", nativeName: "English", locale: "en-IN", direction: "ltr", enabled: true, capabilities: NO_CAPABILITIES },
  { code: "hi", name: "Hindi", nativeName: "हिन्दी", locale: "hi-IN", direction: "ltr", enabled: true, capabilities: NO_CAPABILITIES },
  { code: "bn", name: "Bengali", nativeName: "বাংলা", locale: "bn-IN", direction: "ltr", enabled: true, capabilities: NO_CAPABILITIES },
  { code: "ta", name: "Tamil", nativeName: "தமிழ்", locale: "ta-IN", direction: "ltr", enabled: false, capabilities: NO_CAPABILITIES },
  { code: "te", name: "Telugu", nativeName: "తెలుగు", locale: "te-IN", direction: "ltr", enabled: false, capabilities: NO_CAPABILITIES },
  { code: "mr", name: "Marathi", nativeName: "मराठी", locale: "mr-IN", direction: "ltr", enabled: false, capabilities: NO_CAPABILITIES },
  { code: "gu", name: "Gujarati", nativeName: "ગુજરાતી", locale: "gu-IN", direction: "ltr", enabled: false, capabilities: NO_CAPABILITIES },
  { code: "kn", name: "Kannada", nativeName: "ಕನ್ನಡ", locale: "kn-IN", direction: "ltr", enabled: false, capabilities: NO_CAPABILITIES },
  { code: "ml", name: "Malayalam", nativeName: "മലയാളം", locale: "ml-IN", direction: "ltr", enabled: false, capabilities: NO_CAPABILITIES },
  { code: "or", name: "Odia", nativeName: "ଓଡ଼ିଆ", locale: "or-IN", direction: "ltr", enabled: false, capabilities: NO_CAPABILITIES },
  { code: "pa", name: "Punjabi", nativeName: "ਪੰਜਾਬੀ", locale: "pa-IN", direction: "ltr", enabled: false, capabilities: NO_CAPABILITIES },
  { code: "as", name: "Assamese", nativeName: "অসমীয়া", locale: "as-IN", direction: "ltr", enabled: false, capabilities: NO_CAPABILITIES },
  { code: "ur", name: "Urdu", nativeName: "اردو", locale: "ur-IN", direction: "rtl", enabled: false, capabilities: NO_CAPABILITIES },
];

export const DEFAULT_LANGUAGE: LanguageCode = "en";

const BY_CODE = new Map<string, LanguageDefinition>(LANGUAGES.map((lang) => [lang.code, lang]));

/** One language definition by code, or undefined if unregistered. Never
 * throws -- callers decide whether an unknown code means "fall back to
 * English" or "reject the request". */
export function getLanguage(code: string): LanguageDefinition | undefined {
  return BY_CODE.get(code);
}

/** True only for a language that is both registered AND enabled -- the
 * check almost always wanted together (e.g. validating a persisted
 * language preference before applying it). */
export function isSupported(code: string): boolean {
  const lang = BY_CODE.get(code);
  return lang !== undefined && lang.enabled;
}

/** Every language currently selectable in the UI (Milestone 1:
 * en/hi/bn), in registry order. */
export function enabledLanguages(): LanguageDefinition[] {
  return LANGUAGES.filter((lang) => lang.enabled);
}

/** Every registered code, enabled or not. */
export function supportedCodes(): readonly string[] {
  return LANGUAGES.map((lang) => lang.code);
}

export function isRtl(code: string): boolean {
  return getLanguage(code)?.direction === "rtl";
}
